### PR TITLE
Feature: Add consistent logging errorprone rule

### LINKED ISSUE #1642: Enforce casing and naming of slf4j static `Logger` field
## What happened?

Our tooling currently allows people to name their logger fields anything they want (and also tolerates different capitalizations):
```java
private static final Logger log = LoggerFactory.getLogger(MyClass.class);
private static final Logger LOG = LoggerFactory.getLogger(MyClass.class);
private static final Logger LOGGER = LoggerFactory.getLogger(MyClass.class);
private static final Logger logger = LoggerFactory.getLogger(MyClass.class);
```

I've observed a couple of code-review nit threads where the one of these has been picked and the maintainers would prefer another.

## What did you want to happen?

I'd like tooling to just enforce this (and ideally auto-fix everywhere), because I don't think it's a valuable thing for authors/reviewers to spend time on and reduces the signal to noise ratio of code reviews. My preferred variant would be:

```java
private static final Logger log = LoggerFactory.getLogger(MyClass.class);
```

### PR DESCRIPTION
**Pull request by ferozco**
## Before this PR
Closes #1642 

## After this PR
==COMMIT_MSG==
Add consistent logging errorprone rule
==COMMIT_MSG==

### CODE REVIEW

**Comment by iamdanfox** [comment]
Depending on how controversial this is, we might consider auto-fixing this by adding here https://github.com/palantir/gradle-baseline/blob/928e115f5598644f613e2715a8783742759e95f4/gradle-baseline-java/src/main/groovy/com/palantir/baseline/extensions/BaselineErrorProneExtension.java#L29

File: baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/ConsistentLoggerName.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+@BugPattern(
+        name = "ConsistentLoggerName",
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        providesFix = BugPattern.ProvidesFix.REQUIRES_HUMAN_ATTENTION,
+        severity = BugPattern.SeverityLevel.WARNING,
+        summary = "Loggers created using getLogger(Class<?>) must be named 'log'.")
+public final class ConsistentLoggerName extends BugChecker implements BugChecker.VariableTreeMatcher {
+
+    private static final Matcher<VariableTree> matcher = Matchers.allOf(
```
Comment (carterkozak) [inline baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/ConsistentLoggerName.java:41]: Thoughts on filtering down to `static final` fields in this case? I think we should add other rules to require loggers to be private+static+final, but in the case of odd legacy code we should flag the most specific failure. I'm not confident we'd want instance loggers to be renamed automatically, I just don't have a good enough sense for all the places they might exist.

**Comment by carterkozak** [comment]
+1 for auto-fixing this, please give the automatic fix a run on our large internal codebase to verify before we merge.

File: baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/ConsistentLoggerName.java [hunk 2]
```diff
@@ -48,16 +50,13 @@
 
     @Override
     public Description matchVariable(VariableTree tree, VisitorState state) {
-        if (!matcher.matches(tree, state)) {
-            return Description.NO_MATCH;
+        if (matcher.matches(tree, state)) {
+            if (!tree.getName().contentEquals("log")) {
```
Comment (carterkozak) [inline baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/ConsistentLoggerName.java]: pedantic nit: might as well combine conditions to reduce indent depth `if (matcher.matches(tree, state) && !tree.getName().contentEquals("log")) {`

I'll give this PR a try internally

File: baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/ConsistentLoggerName.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        linkType = BugPattern.LinkType.CUSTOM,
+        providesFix = BugPattern.ProvidesFix.REQUIRES_HUMAN_ATTENTION,
+        severity = BugPattern.SeverityLevel.WARNING,
+        summary = "Loggers created using getLogger(Class<?>) must be named 'log'.")
+public final class ConsistentLoggerName extends BugChecker implements BugChecker.VariableTreeMatcher {
+
+    private static final Matcher<VariableTree> matcher = Matchers.allOf(
+            Matchers.isField(),
+            Matchers.isStatic(),
+            MoreMatchers.isFinal(),
```
Comment (carterkozak) [inline baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/ConsistentLoggerName.java]: I think we can avoid adding a new utility:
```suggestion
            Matchers.hasModifier(Modifier.FINAL),
```

That will match `interface Iface { Logger log = LoggerFactory.getLogger(Iface.class); }` as `final` which is why we had to implement the custom `MoreMatchers.hasExplicitModifier` for our `RedundantModifier` check.

File: baseline-error-prone/src/test/java/com/palantir/baseline/errorprone/ConsistentLoggerNameTest.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+class ConsistentLoggerNameTest {
+
+    @Test
+    void testFix() {
+        fix().addInputLines(
+                        "Test.java",
+                        "import org.slf4j.*;",
+                        "class Test {",
+                        "    private static final Logger LOG = LoggerFactory.getLogger(Test.class);",
```
Comment (uschi2000) [inline baseline-error-prone/src/test/java/com/palantir/baseline/errorprone/ConsistentLoggerNameTest.java:29]: how about `logger` -> `log`? or do you think that's cool?

File: baseline-error-prone/src/test/java/com/palantir/baseline/errorprone/ConsistentLoggerNameTest.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    @Test
+    void ignores_local_variables() {
+        fix().addInputLines(
+                        "Test.java",
+                        "import org.slf4j.*;",
+                        "class Test {",
+                        "    private Logger LOG = LoggerFactory.getLogger(Test.class);",
+                        "}")
+                .addOutputLines(
```
Comment (carterkozak) [inline baseline-error-prone/src/test/java/com/palantir/baseline/errorprone/ConsistentLoggerNameTest.java]: Perhaps `.expectUnchanged()`

File: baseline-error-prone/src/test/java/com/palantir/baseline/errorprone/ConsistentLoggerNameTest.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+    @Test
+    void ignores_non_final_methods() {
+        fix().addInputLines(
+                        "Test.java",
+                        "import org.slf4j.*;",
+                        "class Test {",
+                        "    private static Logger LOG = LoggerFactory.getLogger(Test.class);",
+                        "}")
+                .addOutputLines(
```
Comment (carterkozak) [inline baseline-error-prone/src/test/java/com/palantir/baseline/errorprone/ConsistentLoggerNameTest.java]: `.expectUnchanged()`

File: baseline-error-prone/src/test/java/com/palantir/baseline/errorprone/ConsistentLoggerNameTest.java (hunk 4 above)
Comment (ferozco) [reply baseline-error-prone/src/test/java/com/palantir/baseline/errorprone/ConsistentLoggerNameTest.java:29]: Everything goes to `log`. There can only be 1 name for loggers

**Review by carterkozak — APPROVED** [review]
thanks!

**Pull request merged**